### PR TITLE
Feat: Klaviyo ecommerce Specs

### PR DESCRIPTION
**Pull request by abandonned**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/klaviyo/transform.js [hunk 1]
```diff
@@ -2,13 +2,19 @@
 /* eslint-disable no-underscore-dangle */
 /* eslint-disable  array-callback-return */
 const get = require("get-value");
+const fs = require("fs");
+const path = require("path");
```
Comment (akashrpo) [inline v0/destinations/klaviyo/transform.js:6]: Let's remove these unused imports

File: v0/destinations/klaviyo/transform.js [hunk 2]
```diff
@@ -160,13 +163,75 @@ const trackRequestHandler = (message, category, destination) => {
...
-    customerProperties._id = getFieldValueFromMessage(message, "userId");
-  }
-  payload.customer_properties = customerProperties;
-  if (message.timestamp) {
-    payload.time = toUnixTimestamp(message.timestamp);
+  return customerProperties;
+};
+const trackRequestHandler = (message, category, destination) => {
+  let payload = {};
+  let event = getValueFromMessage(message, "event");
```
Comment (akashrpo) [inline v0/destinations/klaviyo/transform.js:170]: Let's use simple get() call to extract `event`

**Comment by akashrpo** [comment]
Let's add required test cases for new ecomm specs so we don't miss out any edge cases during merge.

File: v0/destinations/klaviyo/transform.js (hunk 2 above)
Comment (abandonned) [reply v0/destinations/klaviyo/transform.js:170]: Everywhere in transform.js getValueFromMessage is being used , so let's maintain parity if there is no advantage in changing that.

Comment (akashrpo) [reply v0/destinations/klaviyo/transform.js:168]: Actually we are using only `get()` and `getFieldValueFromMessage()` in transform.js so directly calling `get()` here would be better.

File: v0/destinations/klaviyo/config.js [hunk 3]
```diff
@@ -9,9 +9,49 @@ const CONFIG_CATEGORIES = {
   IDENTIFY: { name: "KlaviyoIdentify", apiUrl: "/api/identify" },
   SCREEN: { name: "KlaviyoTrack", apiUrl: "/api/track" },
   TRACK: { name: "KlaviyoTrack", apiUrl: "/api/track" },
-  GROUP: { name: "KlaviyoGroup" }
+  GROUP: { name: "KlaviyoGroup" },
+  STARTED_CHECKOUT: { type: "STARTED_CHECKOUT", name: "StartedCheckout" },
+  VIEWED_PRODUCT: { type: "VIEWED_PRODUCT", name: "ViewedProduct" },
+  ADDED_TO_CART: { type: "ADDED_TO_CART", name: "AddedToCart" },
+  ITEMS: { type: "ITEMS", name: "Items" }
```
Comment (saikumarrs) [inline v0/destinations/klaviyo/config.js:16]: Just curious on why `type` was added to only the new entries

File: v0/destinations/klaviyo/transform.js (hunk 2 above)
Comment (saikumarrs) [reply v0/destinations/klaviyo/transform.js:168]: If we need to extract value from a key that does not have multiple sources (say `userId` --- `userId`, `traits.userId`, `traits.id` etc.), then `get` is ideal, as in this case. Otherwise, `getFieldValueFromMessage` should be used.

File: v0/destinations/klaviyo/transform.js [hunk 4]
```diff
@@ -160,13 +161,75 @@ const trackRequestHandler = (message, category, destination) => {
...
+    }
+    const customerProperties = createCustomerProperties(message);
+    if (destination.Config.enforceEmailAsPrimary) {
+      delete customerProperties.$id;
+      customerProperties._id = getFieldValueFromMessage(message, "userId");
+    }
+    payload.customer_properties = customerProperties;
+    if (message.timestamp) {
+      payload.time = toUnixTimestamp(message.timestamp);
+    }
```
Comment (saikumarrs) [inline v0/destinations/klaviyo/transform.js:232]: The entire logic is cramped here with no visual separation or code documentation. Please improve this.

File: v0/destinations/klaviyo/transform.js [hunk 5]
```diff
@@ -160,13 +160,75 @@ const trackRequestHandler = (message, category, destination) => {
...
+    const eventName = eventNameMapping[event];
+    payload.event = eventName;
+    payload.token = destination.Config.publicApiKey;
+    const eventMap = jsonNameMapping[eventName];
+    payload.customer_properties = createCustomerProperties(message);
+    if (
+      !payload.customer_properties.$email &&
+      !payload.customer_properties.$phone_number
+    ) {
+      throw new CustomError("customer_properties not found", 400);
```
Comment (saikumarrs) [inline v0/destinations/klaviyo/transform.js:181]: The error is not appropriate here. Let's make it a bit more specific

File: v0/destinations/klaviyo/transform.js [hunk 6]
```diff
@@ -160,13 +160,75 @@ const trackRequestHandler = (message, category, destination) => {
...
+      delete payload.properties.revenue;
+    }
+    const customerProperties = createCustomerProperties(message);
+    if (destination.Config.enforceEmailAsPrimary) {
+      delete customerProperties.$id;
+      customerProperties._id = getFieldValueFromMessage(message, "userId");
+    }
+    payload.customer_properties = customerProperties;
+    if (message.timestamp) {
+      payload.time = toUnixTimestamp(message.timestamp);
```
Comment (saikumarrs) [inline v0/destinations/klaviyo/transform.js:236]: Why are these not needed for e-com events?

Comment (abandonned) [reply v0/destinations/klaviyo/transform.js:236]: These are pre existing track call. didn't change that, and hence its inside else block

Comment (saikumarrs) [reply v0/destinations/klaviyo/transform.js:236]: So, not applicable for e-com events?

Comment (saikumarrs) [reply v0/destinations/klaviyo/transform.js:236]: Actually, there is a good amount of common logic b/w two categories. I would unify the logic and put the event type-specific code after that (as already there are `if` conditions to perform specific actions)

Comment (abandonned) [reply v0/destinations/klaviyo/transform.js:236]: its not related to ecom events

File: v0/destinations/klaviyo/transform.js (hunk 4 above)
Comment (abandonned) [reply v0/destinations/klaviyo/transform.js]: have added comments and separated different things with linebreak

File: v0/destinations/klaviyo/transform.js (hunk 5 above)
Comment (abandonned) [reply v0/destinations/klaviyo/transform.js]: Have updated the error code to provide more information

File: v0/destinations/klaviyo/transform.js [hunk 7]
```diff
@@ -160,11 +160,85 @@ const trackRequestHandler = (message, category, destination) => {
...
+    const customerProperties = createCustomerProperties(message);
+    if (destination.Config.enforceEmailAsPrimary) {
+      delete customerProperties.$id;
+      customerProperties._id = getFieldValueFromMessage(message, "userId");
+    }
+    payload.customer_properties = customerProperties;
   }
-  payload.customer_properties = customerProperties;
   if (message.timestamp) {
     payload.time = toUnixTimestamp(message.timestamp);
```
Comment (akashrpo) [inline v0/destinations/klaviyo/transform.js:243]: Let's use `getFieldValueFromMessage()` to extract timestamp and also insert a check whether timestamp is already in unix format otherwise this function will truncate the last 3 digits and contain invalid data.

[the remark this answers is not in the document]
Comment (abandonned) [reply v0/destinations/klaviyo/transform.js:243]: this part of the code is already in production.  @saikumarrs should we update it or keep the original one?

**Review by akashrpo — APPROVED** [review]

File: v0/destinations/klaviyo/transform.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (saikumarrs) [reply v0/destinations/klaviyo/transform.js:243]: I wouldn't go for this update as `timestamp` and `originalTimestamp` by definition are not same. 
Refer to https://rudderstack.com/docs/data-warehouse-integrations/warehouse-schemas/#standard-rudderstack-properties.

For the second concern, as both the timestamp fields are calculated/generated by RudderStack and wouldn't be in a different format than expected.

**Review by saikumarrs — APPROVED** [review]
